File: GPM2.0/app/frontend/src/dev-bridge/operations/assembly.js
import {
  normalizeToken,
  parseKeyValueLines,
  requireNumber,
  requireString,
} from "../contracts.js";

export function createAssemblyOperations(runtime) {
  const { runBackend } = runtime;

async function listChrViewCtgs(payload) {
  const { workspaceRoot, projectId, chrName, datasetId = null } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireString("chrName", chrName);

  const args = [
    "list-chr-view-ctgs",
    workspaceRoot,
    String(projectId),
    "--chr-name",
    chrName,
  ];
  if (Number.isFinite(Number(datasetId)) && Number(datasetId) > 0) {
    args.push("--dataset-id", String(Math.trunc(Number(datasetId))));
  }
  const output = await runBackend(args);
  const lines = output.stdout.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const items = [];
  for (const line of lines) {
    const match = line.match(
      /^ctg id=(\d+) name=(.*?) chr=(.*?) chr_order=(\S+) anchor_start=(\S+) ref_orient=(\S+) mode=(\S+) members=(\d+) bp=(\d+)(?: dataset_id=(\S+) dataset=(.*))?$/,
    );
    if (!match) {
      continue;
    }
    items.push({
      assemblyCtgId: Number(match[1]),
      name: match[2],
      assignedChrName: normalizeToken(match[3]),
      chrOrder: normalizeToken(match[4]),
      anchorStart: normalizeToken(match[5]),
      refOrient: normalizeToken(match[6]),
      placementMode: match[7],
      memberCount: Number(match[8]),
      totalLength: Number(match[9]),
      datasetId: normalizeToken(match[10]),
      datasetName: normalizeToken(match[11]),
    });
  }
  return { items };
}

async function listReferenceTrackMembers(payload) {
  const { workspaceRoot, projectId, chrName } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireString("chrName", chrName);

  const output = await runBackend([
    "list-reference-track-members",
    workspaceRoot,
    String(projectId),
    chrName,
  ]);
  const lines = output.stdout.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const items = [];
  const memberByOrder = new Map();

  for (const line of lines) {
    const memberMatch = line.match(
      /^member order=(\d+) ref_chr_id=(\d+) name=(.*?) chr=(.*?) start=(\d+) end=(\d+) anchor_start=(\d+) ref_orient=(\S+) bp=(\d+) hits=(\d+)$/,
    );
    if (memberMatch) {
      const item = {
        sourceKind: "ref_segment",
        referenceChrId: Number(memberMatch[2]),
        referenceChrName: memberMatch[4],
        segmentOrder: Number(memberMatch[1]),
        segmentStartBp: Number(memberMatch[5]),
        segmentEndBp: Number(memberMatch[6]),
        name: memberMatch[3],
        anchorStart: Number(memberMatch[7]),
        totalLength: Number(memberMatch[9]),
        refOrient: memberMatch[8],
        hits: [],
      };
      items.push(item);
      memberByOrder.set(item.segmentOrder, item);
      continue;
    }

    const hitMatch = line.match(
      /^hit member_order=(\d+) hit_id=(\d+) dataset_id=(\d+) source_seq_id=(\d+) strand=(\S+) query_start=(\d+) query_end=(\d+) ref_start=(\d+) ref_end=(\d+) block_length=(\d+) mapq=(\d+) ctg_start=(\d+) ctg_end=(\d+)$/,
    );
    if (!hitMatch) {
      continue;
    }
    const parent = memberByOrder.get(Number(hitMatch[1]));
    if (!parent) {
      continue;
    }
    parent.hits.push({
      hitId: Number(hitMatch[2]),
      datasetId: Number(hitMatch[3]),
      sourceSeqId: Number(hitMatch[4]),
      strand: hitMatch[5],
      queryStart: Number(hitMatch[6]),
      queryEnd: Number(hitMatch[7]),
      refStart: Number(hitMatch[8]),
      refEnd: Number(hitMatch[9]),
      blockLength: Number(hitMatch[10]),
      mapq: Number(hitMatch[11]),
      ctgStart: Number(hitMatch[12]),
      ctgEnd: Number(hitMatch[13]),
    });
  }

  return { items };
}

async function listPhasedChrTracks(payload) {
  const { workspaceRoot, projectId, parentChrName } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireString("parentChrName", parentChrName);
  const output = await runBackend([
    "list-phased-chr-tracks",
    workspaceRoot,
    String(projectId),
    parentChrName,
  ]);
  return parsePhasedChrTracks(output.stdout, projectId, parentChrName);
}

async function createPhasedChrTrack(payload) {
  const { workspaceRoot, projectId, parentChrName } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireString("parentChrName", parentChrName);
  const output = await runBackend([
    "create-phased-chr-track",
    workspaceRoot,
    String(projectId),
    parentChrName,
  ]);
  const record = parseKeyValueLines(output.stdout);
  return {
    track: {
      phasedTrackId: Number(record.phased_track_id || 0),
      projectId: Number(record.project_id || projectId),
      parentChrName: record.parent_chr_name || parentChrName,
      haplotypeKey: record.haplotype_key || "",
      label: record.label || "",
      displayOrder: Number(record.display_order || 0),
      items: [],
    },
  };
}

async function deletePhasedChrTrack(payload) {
  const { workspaceRoot, projectId, phasedTrackId } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireNumber("phasedTrackId", phasedTrackId);
  const output = await runBackend([
    "delete-phased-chr-track",
    workspaceRoot,
    String(projectId),
    String(phasedTrackId),
  ]);
  const record = parseKeyValueLines(output.stdout);
  return {
    projectId: Number(record.project_id || projectId),
    phasedTrackId: Number(record.phased_track_id || phasedTrackId),
    parentChrName: record.parent_chr_name || "",
    haplotypeKey: record.haplotype_key || "",
    label: record.label || "",
    deleted: record.deleted === "true",
  };
}

async function addCtgToPhasedChrTrack(payload) {
  const { workspaceRoot, projectId, phasedTrackId, assemblyCtgId } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireNumber("phasedTrackId", phasedTrackId);
  requireNumber("assemblyCtgId", assemblyCtgId);
  const output = await runBackend([
    "add-ctg-to-phased-chr-track",
    workspaceRoot,
    String(projectId),
    String(phasedTrackId),
    String(assemblyCtgId),
  ]);
  const record = parseKeyValueLines(output.stdout);
  return {
    item: {
      itemId: Number(record.phased_track_item_id || 0),
      phasedTrackId: Number(record.phased_track_id || phasedTrackId),
      assemblyCtgId: Number(record.assembly_ctg_id || assemblyCtgId),
      displayOrder: Number(record.display_order || 0),
      gapBeforePx: Number(record.gap_before_px || 0),
    },
  };
}

async function removePhasedChrTrackItem(payload) {
  const { workspaceRoot, projectId, phasedTrackItemId } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireNumber("phasedTrackItemId", phasedTrackItemId);
  const output = await runBackend([
    "remove-phased-chr-track-item",
    workspaceRoot,
    String(projectId),
    String(phasedTrackItemId),
  ]);
  const record = parseKeyValueLines(output.stdout);
  return {
    projectId: Number(record.project_id || projectId),
    phasedTrackId: Number(record.phased_track_id || 0),
    phasedTrackItemId: Number(record.phased_track_item_id || phasedTrackItemId),
    removed: record.removed === "true",
  };
}

async function reorderPhasedChrTrackItems(payload) {
  const { workspaceRoot, projectId, phasedTrackId, itemIds } = payload || {};
  requireString("workspaceRoot", workspaceRoot);
  requireNumber("projectId", projectId);
  requireNumber("phasedTrackId", phasedTrackId);
  if (!Array.isArray(itemIds)) {
    throw new Error("itemIds must be an array");
  }
  const output = await runBackend([
    "reorder-phased-chr-track-items",
    workspaceRoot,
    String(projectId),
    String(phasedTrackId),
    itemIds.join(","),
  ]);
  const record = parseKeyValueLines(output.stdout);
  return {
    projectId: Number(record.project_id || projectId),
    phasedTrackId: Number(record.phased_track_id || phasedTrackId),
    itemCount: Number(record.item_count || 0),
  };
}

function parsePhasedChrTracks(stdout, fallbackProjectId, fallbackParentChrName) {
  const lines = stdout.split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const record = parseKeyValueLines(stdout);
  const tracks = [];
  let currentTrack = null;
  for (const line of lines) {
    const trackMatch = line.match(
      /^track id=(\d+) parent_chr_name=(.*?) haplotype_key=(.*?) label=(.*?) display_order=(\d+) item_count=(\d+)$/,
    );
    if (trackMatch) {
      currentTrack = {
        phasedTrackId: Number(trackMatch[1]),
        projectId: Number(record.project_id || fallbackProjectId),
        parentChrName: trackMatch[2],
        haplotypeKey: trackMatch[3],
        label: trackMatch[4],
        displayOrder: Number(trackMatch[5]),
        items: [],
      };
      tracks.push(currentTrack);
      continue;
    }

    const itemMatch = line.match(
      /^item id=(\d+) phased_track_id=(\d+) assembly_ctg_id=(\d+) display_order=(\d+) gap_before_px=(\d+)$/,
    );
    if (!itemMatch || !currentTrack) {
      continue;
    }
    currentTrack.items.push({
      itemId: Number(itemMatch[1]),
      phasedTrackId: Number(itemMatch[2]),
      assemblyCtgId: Number(itemMatch[3]),
      displayOrder: Number(itemMatch[4]),
      gapBeforePx: Number(itemMatch[5]),
    });
  }
  return {
    projectId: Number(record.project_id || fallbackProjectId),
    parentChrName: record.parent_chr_name || fallbackParentChrName,
    tracks,
  };
}

  return {
    listChrViewCtgs,
    listReferenceTrackMembers,
    listPhasedChrTracks,
    createPhasedChrTrack,
    deletePhasedChrTrack,
    addCtgToPhasedChrTrack,
    removePhasedChrTrackItem,
    reorderPhasedChrTrackItems,
  };
}

File: GPM2.0/app/backend/src/cli/assembly.rs
use super::*;

pub(super) fn dispatch(command: Commands) -> Result<Option<Commands>> {
    match command {
        Commands::ListChrViewCtgs {
            workspace_root,
            project_id,
            chr_name,
            dataset_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let items = list_chr_view_ctgs(
                &project_db_path,
                project_id,
                chr_name.as_deref(),
                dataset_id,
            )?;
            println!("ctg_count={}", items.len());
            for item in items {
                println!(
                    "ctg id={} name={} chr={} chr_order={} anchor_start={} ref_orient={} mode={} members={} bp={} dataset_id={} dataset={}",
                    item.assembly_ctg_id,
                    item.name,
                    item.assigned_chr_name
                        .unwrap_or_else(|| "UNPLACED".to_string()),
                    opt_i64(item.chr_order),
                    opt_i64(item.anchor_start),
                    item.ref_orient.unwrap_or_else(|| "NULL".to_string()),
                    item.placement_mode,
                    item.member_count,
                    item.total_length,
                    opt_i64(item.dataset_id),
                    item.dataset_name.as_deref().unwrap_or("MIXED")
                );
            }
        }
        Commands::ListReferenceTrackMembers {
            workspace_root,
            project_id,
            chr_name,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let items = list_reference_track_members(&project_db_path, project_id, &chr_name)?;
            println!("member_count={}", items.len());
            for item in items {
                println!(
                    "member order={} ref_chr_id={} name={} chr={} start={} end={} anchor_start={} ref_orient={} bp={} hits={}",
                    item.segment_order,
                    item.reference_chr_id,
                    item.name,
                    item.reference_chr_name,
                    item.segment_start_bp,
                    item.segment_end_bp,
                    item.anchor_start,
                    item.ref_orient,
                    item.total_length,
                    item.hits.len()
                );
                for hit in item.hits {
                    println!(
                        "hit member_order={} hit_id={} dataset_id={} source_seq_id={} strand={} query_start={} query_end={} ref_start={} ref_end={} block_length={} mapq={} ctg_start={} ctg_end={}",
                        item.segment_order,
                        hit.hit_id,
                        hit.dataset_id,
                        hit.source_seq_id,
                        hit.strand,
                        hit.query_start,
                        hit.query_end,
                        hit.ref_start,
                        hit.ref_end,
                        hit.block_length,
                        hit.mapq,
                        hit.ctg_start,
                        hit.ctg_end
                    );
                }
            }
        }
        Commands::ListPhasedChrTracks {
            workspace_root,
            project_id,
            parent_chr_name,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let result = list_phased_chr_tracks(&project_db_path, project_id, &parent_chr_name)?;
            print_phased_chr_tracks(&result);
        }
        Commands::CreatePhasedChrTrack {
            workspace_root,
            project_id,
            parent_chr_name,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = create_phased_chr_track(&project_db_path, project_id, &parent_chr_name)?;
            println!("project_id={}", summary.track.project_id);
            println!("phased_track_id={}", summary.track.id);
            println!("parent_chr_name={}", summary.track.parent_chr_name);
            println!("haplotype_key={}", summary.track.haplotype_key);
            println!("label={}", summary.track.label);
            println!("display_order={}", summary.track.display_order);
            println!("item_count={}", summary.track.items.len());
        }
        Commands::DeletePhasedChrTrack {
            workspace_root,
            project_id,
            phased_track_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = delete_phased_chr_track(&project_db_path, project_id, phased_track_id)?;
            println!("project_id={}", summary.project_id);
            println!("phased_track_id={}", summary.phased_track_id);
            println!("parent_chr_name={}", summary.parent_chr_name);
            println!("haplotype_key={}", summary.haplotype_key);
            println!("label={}", summary.label);
            println!("deleted={}", summary.deleted);
        }
        Commands::AddCtgToPhasedChrTrack {
            workspace_root,
            project_id,
            phased_track_id,
            assembly_ctg_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = add_ctg_to_phased_chr_track(
                &project_db_path,
                project_id,
                phased_track_id,
                assembly_ctg_id,
            )?;
            let item = summary.item;
            println!("project_id={}", project_id);
            println!("phased_track_item_id={}", item.id);
            println!("phased_track_id={}", item.phased_track_id);
            println!("assembly_ctg_id={}", item.assembly_ctg_id);
            println!("display_order={}", item.display_order);
            println!("gap_before_px={}", item.gap_before_px);
            println!("orient={}", item.orient);
        }
        Commands::RemovePhasedChrTrackItem {
            workspace_root,
            project_id,
            phased_track_item_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary =
                remove_phased_chr_track_item(&project_db_path, project_id, phased_track_item_id)?;
            println!("project_id={}", summary.project_id);
            println!("phased_track_id={}", summary.phased_track_id);
            println!("phased_track_item_id={}", summary.phased_track_item_id);
            println!("removed={}", summary.removed);
        }
        Commands::ReorderPhasedChrTrackItems {
            workspace_root,
            project_id,
            phased_track_id,
            item_ids,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let item_ids = parse_id_list(Some(item_ids))?;
            let summary = reorder_phased_chr_track_items(
                &project_db_path,
                project_id,
                phased_track_id,
                &item_ids,
            )?;
            println!("project_id={}", summary.project_id);
            println!("phased_track_id={}", summary.phased_track_id);
            println!("item_count={}", summary.item_count);
        }
        Commands::GetCtgDetail {
            workspace_root,
            project_id,
            assembly_ctg_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let detail = get_ctg_detail(&project_db_path, project_id, assembly_ctg_id)?;
            println!(
                "ctg id={} name={} chr={} chr_order={} anchor_start={} ref_orient={} mode={}",
                detail.assembly_ctg_id,
                detail.name,
                detail
                    .assigned_chr_name
                    .unwrap_or_else(|| "UNPLACED".to_string()),
                opt_i64(detail.chr_order),
                opt_i64(detail.anchor_start),
                detail.ref_orient.unwrap_or_else(|| "NULL".to_string()),
                detail.placement_mode
            );
            println!("member_count={}", detail.members.len());
            for member in detail.members {
                println!(
                    "member id={} order={} assembly_seq_id={} dataset={} seq={} len={} orient={} range={}..{} left_end_type={} right_end_type={} hidden={}",
                    member.assembly_ctg_member_id,
                    member.member_order,
                    member.assembly_seq_id,
                    member.dataset_name,
                    member.seq_name,
                    member.seq_length,
                    member.orient,
                    member.source_start,
                    member.source_end,
                    member.left_end_type,
                    member.right_end_type,
                    member.hidden
                );
            }
        }
        Commands::ListCtgEditCandidates {
            workspace_root,
            project_id,
            assembly_ctg_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let candidates =
                list_ctg_edit_candidates(&project_db_path, project_id, assembly_ctg_id)?;
            println!(
                "move_target_count={} add_seq_count={}",
                candidates.move_target_ctgs.len(),
                candidates.add_seq_candidates.len()
            );
            for ctg in candidates.move_target_ctgs {
                println!(
                    "move_target id={} name={} chr={} chr_order={}",
                    ctg.assembly_ctg_id,
                    ctg.name,
                    ctg.assigned_chr_name
                        .unwrap_or_else(|| "UNPLACED".to_string()),
                    opt_i64(ctg.chr_order)
                );
            }
            for seq in candidates.add_seq_candidates {
                println!(
                    "add_seq assembly_seq_id={} dataset={} seq={} len={} hidden={}",
                    seq.assembly_seq_id, seq.dataset_name, seq.seq_name, seq.seq_length, seq.hidden
                );
            }
        }
        Commands::BootstrapProjectAssembly {
            workspace_root,
            project_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = bootstrap_project_assembly(&project_db_path, project_id)?;
            println!("project_id={}", summary.project_id);
            println!("assembly_seq_count={}", summary.assembly_seq_count);
            println!("assembly_ctg_count={}", summary.assembly_ctg_count);
            println!("assembly_member_count={}", summary.assembly_member_count);
        }
        Commands::AutoAssignChr {
            workspace_root,
            project_id,
            alignment_block_size,
            alignment_coverage_percent,
            assign_unplaced,
            reposition_anchored,
            skip_manual,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = auto_assign_chr(
                &project_db_path,
                project_id,
                &AutoAssignChrParams {
                    alignment_block_size,
                    alignment_coverage_percent,
                    assign_unplaced,
                    reposition_anchored,
                    skip_manual,
                },
            )?;
            println!("project_id={}", summary.project_id);
            println!("processed_ctg_count={}", summary.processed_ctg_count);
            println!("assigned_count={}", summary.assigned_count);
            println!("repositioned_count={}", summary.repositioned_count);
            println!("cleared_count={}", summary.cleared_count);
            println!("skipped_manual_count={}", summary.skipped_manual_count);
            println!("no_evidence_count={}", summary.no_evidence_count);
            println!("refreshed_chr_count={}", summary.refreshed_chr_count);
            println!(
                "loaded_alignment_dataset_count={}",
                summary.loaded_alignment_dataset_count
            );
            println!(
                "loaded_alignment_hit_count={}",
                summary.loaded_alignment_hit_count
            );
        }
        Commands::AutoOrientContigs {
            workspace_root,
            project_id,
            alignment_block_size,
            alignment_coverage_percent,
            skip_manual,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = auto_orient_contigs(
                &project_db_path,
                project_id,
                &AutoOrientContigsParams {
                    alignment_block_size,
                    alignment_coverage_percent,
                    skip_manual,
                },
            )?;
            println!("project_id={}", summary.project_id);
            println!("processed_ctg_count={}", summary.processed_ctg_count);
            println!("oriented_ctg_count={}", summary.oriented_ctg_count);
            println!("flipped_ctg_count={}", summary.flipped_ctg_count);
            println!("no_evidence_count={}", summary.no_evidence_count);
            println!("skipped_manual_count={}", summary.skipped_manual_count);
            println!(
                "loaded_alignment_dataset_count={}",
                summary.loaded_alignment_dataset_count
            );
            println!(
                "loaded_alignment_hit_count={}",
                summary.loaded_alignment_hit_count
            );
        }
        Commands::RenameCtg {
            workspace_root,
            project_id,
            assembly_ctg_id,
            new_name,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = rename_ctg(
                &project_db_path,
                project_id,
                &RenameCtgParams {
                    assembly_ctg_id,
                    new_name,
                },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("old_name={}", summary.old_name);
            println!("new_name={}", summary.new_name);
            println!("changed={}", summary.changed);
        }
        Commands::FlipCtg {
            workspace_root,
            project_id,
            assembly_ctg_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = flip_ctg(
                &project_db_path,
                project_id,
                &FlipCtgParams { assembly_ctg_id },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("member_count={}", summary.member_count);
            println!(
                "ref_orient={}",
                summary.ref_orient.unwrap_or_else(|| "NULL".to_string())
            );
        }
        Commands::FlipSeq {
            workspace_root,
            project_id,
            assembly_seq_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = flip_seq(
                &project_db_path,
                project_id,
                &FlipSeqParams { assembly_seq_id },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_seq_id={}", summary.assembly_seq_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("orient={}", summary.orient);
            println!("left_end_type={}", summary.left_end_type);
            println!("right_end_type={}", summary.right_end_type);
            println!(
                "ref_orient={}",
                summary.ref_orient.unwrap_or_else(|| "NULL".to_string())
            );
        }
        Commands::SetEndType {
            workspace_root,
            project_id,
            assembly_seq_id,
            left_end_type,
            right_end_type,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = set_end_type(
                &project_db_path,
                project_id,
                &SetEndTypeParams {
                    assembly_seq_id,
                    left_end_type,
                    right_end_type,
                },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_seq_id={}", summary.assembly_seq_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("left_end_type={}", summary.left_end_type);
            println!("right_end_type={}", summary.right_end_type);
            println!(
                "ref_orient={}",
                summary.ref_orient.unwrap_or_else(|| "NULL".to_string())
            );
            println!("changed={}", summary.changed);
        }
        Commands::HideSeq {
            workspace_root,
            project_id,
            assembly_seq_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = hide_seq(
                &project_db_path,
                project_id,
                &HideSeqParams { assembly_seq_id },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_seq_id={}", summary.assembly_seq_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("hidden={}", summary.hidden);
            println!(
                "ref_orient={}",
                summary.ref_orient.unwrap_or_else(|| "NULL".to_string())
            );
            println!("changed={}", summary.changed);
        }
        Commands::ShowSeq {
            workspace_root,
            project_id,
            assembly_seq_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = show_seq(
                &project_db_path,
                project_id,
                &ShowSeqParams { assembly_seq_id },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_seq_id={}", summary.assembly_seq_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("hidden={}", summary.hidden);
            println!(
                "ref_orient={}",
                summary.ref_orient.unwrap_or_else(|| "NULL".to_string())
            );
            println!("changed={}", summary.changed);
        }
        Commands::DeleteCtg {
            workspace_root,
            project_id,
            assembly_ctg_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = delete_ctg(
                &project_db_path,
                project_id,
                &DeleteCtgParams { assembly_ctg_id },
            )?;
            println!("project_id={}", summary.project_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!(
                "released_assembly_seq_ids={}",
                summary
                    .released_assembly_seq_ids
                    .iter()
                    .map(|value| value.to_string())
                    .collect::<Vec<_>>()
                    .join(",")
            );
            println!(
                "released_assembly_seq_count={}",
                summary.released_assembly_seq_count
            );
            println!("refreshed_chr_count={}", summary.refreshed_chr_count);
        }
        Commands::ListDeletedCtgs {
            workspace_root,
            project_id,
            chr_name,
            dataset_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let items = list_deleted_ctgs(
                &project_db_path,
                project_id,
                chr_name.as_deref(),
                dataset_id,
            )?;
            println!("project_id={}", project_id);
            println!("deleted_ctg_count={}", items.len());
            for item in items {
                println!(
                    "deleted_ctg record_id={} ctg_id={} name={} chr={} chr_order={} anchor_start={} ref_orient={} mode={} members={} bp={} deleted_at={}",
                    item.deleted_ctg_record_id,
                    item.assembly_ctg_id,
                    item.name,
                    item.assigned_chr_name.unwrap_or_else(|| "NULL".to_string()),
                    opt_i64(item.chr_order),
                    opt_i64(item.anchor_start),
                    item.ref_orient.unwrap_or_else(|| "NULL".to_string()),
                    item.placement_mode,
                    item.member_count,
                    item.total_length,
                    item.deleted_at
                );
            }
        }
        Commands::RestoreDeletedCtg {
            workspace_root,
            project_id,
            deleted_ctg_record_id,
        } => {
            let project_db_path = workspace_root.join("project.sqlite");
            let summary = restore_deleted_ctg(
                &project_db_path,
                project_id,
                &RestoreDeletedCtgParams {
                    deleted_ctg_record_id,
                },
            )?;
            println!("project_id={}", summary.project_id);
            println!("deleted_ctg_record_id={}", summary.deleted_ctg_record_id);
            println!("assembly_ctg_id={}", summary.assembly_ctg_id);
            println!("restored_member_count={}", summary.restored_member_count);
            println!("refreshed_chr_count={}", summary.refreshed_chr_count);
        }
        command => return Ok(Some(command)),
    }
    Ok(None)
}
